src/network/network_client.cpp

```diff
@@ -577,19 +581,29 @@ NetworkRecvStatus ClientNetworkGameSocketHandler::Receive_SERVER_BANNED(Packet *
 
 NetworkRecvStatus ClientNetworkGameSocketHandler::Receive_SERVER_GAME_INFO(Packet *p)
 {
-	if (this->status != STATUS_COMPANY_INFO && this->status != STATUS_INACTIVE) return NETWORK_RECV_STATUS_MALFORMED_PACKET;
+	if (this->status != STATUS_COMPANY_INFO && this->status != STATUS_GAME_INFO) return NETWORK_RECV_STATUS_MALFORMED_PACKET;
+
+	/* Retrieve the NetworkGameInfo from the packet. */
+	NetworkGameInfo ngi = {};
+	DeserializeNetworkGameInfo(p, &ngi);
 
 	NetworkGameList *item = GetLobbyGameInfo();
+	if (item == nullptr) {
+		/* This is not the lobby, so add it to the game list. */
+		item = NetworkGameListAddItem(this->address);
+	}
 
 	/* Clear any existing GRFConfig chain. */
 	ClearGRFConfigList(&item->info.grfconfig);
-	/* Retrieve the NetworkGameInfo from the packet. */
-	DeserializeNetworkGameInfo(p, &item->info);
+	/* Copy the new NetworkGameInfo info. */
+	memcpy(&item->info, &ngi, sizeof(item->info));
```

Comment:
I'm not seeing the benefit of deserializing the NetworkGameInfo about 10 lines before, and then copying it here. Why won't the old deserialization logic work? Even then `item->info = ngi;` seems the better solution as that keeps things working when the fixed C-string buffers are replaced with std::string or some other C++-ish thing is added to that struct.